in-progress/7025-instrumenting-the-node-with-open-telemetry.md

```diff
@@ -0,0 +1,176 @@
+        exporter: new ConsoleMetricExporter(),
+      }),
+      instrumentations: [getNodeAutoInstrumentations()],
+    });
+
+    sdk.start();
+    // ...
+  }
+
+  public stop() {
+    await this.sdk.stop();
+    // ...
+  }
+}
+```
+
+The `getNodeAutoInstrumentations` function from [@opentelemetry/auto-instrumentations-node](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node) sets up instrumentation for a [large number of external packages](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/metapackages/auto-instrumentations-node#supported-instrumentations), the vast majority of which are not being used by Aztec. It will be replaced with smaller set of instrumentations that's relevant to Aztec
+
+> [!NOTE]
+> Starting the SDK will set up the global instances accessible through the `@opentelemetry/api` package so that even external packages can use the backend to expose metrics.
+
+### System metrics
+
+The SDK will be initialized with the [`@opentelemetry/host-metrics`](https://www.npmjs.com/package/@opentelemetry/host-metrics) package to track system resources.
+
+### Exporting data
+
+The main difference between OpenTelemetry and using Prometheus directly is that OpenTelemetry uses a push model (where services push metrics directly to a service/collector), whereas Prometheus uses a pull model (where the prometheus instance scrapes each service individually). Integrating OpenTelemetry would require us to run one extra service in the stack, the OpenTelemetry Collector, that will collect and batch metrics from nodes.
+
+The Sandbox docker compose file will be updated to include an optional OTel connector, Prometheus instance and Grafana dashboard. They will be turned on using compose profiles
+
+```sh
+docker compose up --profile metrics
+```
+
+The Terraform IaC will also be updated to deploy an OTel collector, a Prometheus instance and a Grafana dashboard.
+
+The Grafana dashboard will contain the benchmark stats and system resource utilization for a node. The dashboard will be stored in the repo as code so it's easy to maintain and extend (instead of manually modifying the dashboard in the GUI).
+
+### Existing benchmark tests
+
+The existing benchmark tests will remain unmodified for now. Ideally we'd find a way to capture the metrics data during test runs and recreate the Markdown comment without the need for logging benchmark metrics to stdout. Given the pluggable architecture of OpenTelemetry we should be able to initialize the SDK with a custom `MetricsExporter` that sends the data directly to the existing ndjson files. This would enable us to remove the dependency on `winston`.
+
+## Change Set
+
+Fill in bullets for each area that will be affected by this change.
+
+- [ ] L1 Contracts
+- [ ] Enshrined L2 Contracts
+- [ ] Private Kernel Circuits
+- [ ] Public Kernel Circuits
+- [ ] Rollup Circuits
+- [ ] Aztec.nr
+- [ ] Noir
+- [ ] AVM
+- [x] Sequencer
+- [ ] Fees
+- [ ] P2P Network
```

Comment:
Can we plug the metrics into the P2P module? Would be great if we can.

Reply:
I'll look into getting some stats out of it